File: Medium/ContainerWithMostWater.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int max_area = 0, base = height.size() - 1;

        // start from left most and right most indexes
        int left = 0, right = base;
        while (left < right)
        {
            // move the left or right index based on which side is the smallest
            if (height[left] < height[right])
            {
                max_area = max(max_area, height[left] * base);
                left++;
            }
            else
            {
                max_area = max(max_area, height[right] * base);
                right--;
            }
            base--;
        }
        return max_area;
    }
};

int main()
{
    Solution *s = new Solution();
    vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    cout << s->maxArea(height);
    return 0;
}

File: Python/TwoPointers/ContainerWithMostWater.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        area, base = 0, len(height) - 1
        left, right = 0, base

        while left < right:
            # move left or right pointer depends on which side is smaller
            if (height[left] < height[right]):
                area = max(area, height[left] * base)
                left += 1
            else:
                area = max(area, height[right] * base)
                right -= 1
            base -= 1
        
        return area
    
s = Solution()

s.maxArea([1,8,6,2,5,4,8,3,7])